### ISSUE TITLE
Massive read latency under write load with root on ZFS 0.8-on-LUKS on SSD

### ISSUE THREAD

**Issue by deliciouslytyped** [openzfs/zfs#9148]
TL;DR: switching applications (i.e. alt tab for example) becomes very slow when I put my root (?) dataset under write load.

Longer:

**The Problem**
When I load the drive with writes using `cat /dev/urandom | pv | dd if=/dev/stdin of=./test bs=1M
` the system becomes functionally unusable (this is a simpler test case than `zfs recv`), though the throughput of the command is fine at an acceptable 100MB/s.

I don't know how to use `fio` properly, but two ad-hoc tests I ran for 20 seconds yielded the results below:
```
unloaded:      lat (usec): min=31, max=705, avg=123.25, stdev=54.53
loaded:      lat (usec): min=208, max=39662, avg=14599.32, stdev=12436.19
```
A latency increase from a hundred or so microseconds so many seconds can be clearly seen.

I was unable to experiment with `ionice` given that that only nices the userspace process and not the kernel threads that do the real IO.

**System Info**
I couldn't figure out and good and meaningful reproducible test cases, I would be happy if people could suggest improvements to my methodology.

My root filesystem is on a ZFS dataset which is backed by an encrypted LUKS1 partition. The LUKS partition takes up most of the drive. This LUKS partition is used for the single vdev.
```
$ lsblk
NAME                                          MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda                                             8:0    0 945.5G  0 disk  
├─sda1                                          8:1    0   500M  0 part
├─sda2                                          8:2    0  15G  0 part
└─sda3                                          8:3    0 930G  0 part
  └─root                                      254:0    0 930G  0 crypt
```

The backing hardware is a Samsung 850 Pro which appears to perform fine under normal circumstances. I have 16GB of RAM with no swap, and the system load and memory pressure are both low.

To the best of my knowledge I didn't tweak any ZFS settings.
uname: `Linux nixos 4.19.52 #1-NixOS SMP Mon Jun 17 17:51:57 UTC 2019 x86_64 GNU/Linux
`
```
$ zfs version
zfs-0.8.1-1
zfs-kmod-0.8.1-1
$ modinfo zfs | grep version
version:        0.8.1-1
srcversion:     A253AC98C8B411B821150CC
```
```
$ modinfo spl | grep version
version:        0.8.1-1
srcversion:     F451462905AB5E46555E0E4
```

**Comment by deliciouslytyped** [comment]
Here is `zfs get all` for the root dataset:
```
tank/global1/root                  type                  filesystem                     -
tank/global1/root                  creation              Thu Jul 25 17:14 2019          -
tank/global1/root                  used                  92.5G                          -
tank/global1/root                  available             117G                           -
tank/global1/root                  referenced            92.4G                          -
tank/global1/root                  compressratio         1.00x                          -
tank/global1/root                  mounted               yes                            -
tank/global1/root                  quota                 none                           default
tank/global1/root                  reservation           none                           default
tank/global1/root                  recordsize            128K                           default
tank/global1/root                  mountpoint            legacy                         local
tank/global1/root                  sharenfs              off                            default
tank/global1/root                  checksum              on                             default
tank/global1/root                  compression           off                            default
tank/global1/root                  atime                 on                             default
tank/global1/root                  devices               on                             default
tank/global1/root                  exec                  on                             default
tank/global1/root                  setuid                on                             default
tank/global1/root                  readonly              off                            default
tank/global1/root                  zoned                 off                            default
tank/global1/root                  snapdir               hidden                         default
tank/global1/root                  aclinherit            restricted                     default
tank/global1/root                  createtxg             12                             -
tank/global1/root                  canmount              on                             default
tank/global1/root                  xattr                 on                             default
tank/global1/root                  copies                1                              default
tank/global1/root                  version               5                              -
tank/global1/root                  utf8only              off                            -
tank/global1/root                  normalization         none                           -
tank/global1/root                  casesensitivity       sensitive                      -
tank/global1/root                  vscan                 off                            default
tank/global1/root                  nbmand                off                            default
tank/global1/root                  sharesmb              off                            default
tank/global1/root                  refquota              none                           default
tank/global1/root                  refreservation        none                           default
tank/global1/root                  guid                  [censored]            -
tank/global1/root                  primarycache          all                            default
tank/global1/root                  secondarycache        all                            default
tank/global1/root                  usedbysnapshots       55.4M                          -
tank/global1/root                  usedbydataset         92.4G                          -
tank/global1/root                  usedbychildren        0B                             -
tank/global1/root                  usedbyrefreservation  0B                             -
tank/global1/root                  logbias               latency                        default
tank/global1/root                  objsetid              387                            -
tank/global1/root                  dedup                 off                            default
tank/global1/root                  mlslabel              none                           default
tank/global1/root                  sync                  standard                       default
tank/global1/root                  dnodesize             legacy                         default
tank/global1/root                  refcompressratio      1.00x                          -
tank/global1/root                  written               88.5G                          -
tank/global1/root                  logicalused           90.1G                          -
tank/global1/root                  logicalreferenced     90.1G                          -
tank/global1/root                  volmode               default                        default
tank/global1/root                  filesystem_limit      none                           default
tank/global1/root                  snapshot_limit        none                           default
tank/global1/root                  filesystem_count      none                           default
tank/global1/root                  snapshot_count        none                           default
tank/global1/root                  snapdev               hidden                         default
tank/global1/root                  acltype               off                            default
tank/global1/root                  context               none                           default
tank/global1/root                  fscontext             none                           default
tank/global1/root                  defcontext            none                           default
tank/global1/root                  rootcontext           none                           default
tank/global1/root                  relatime              on                             temporary
tank/global1/root                  redundant_metadata    all                            default
tank/global1/root                  overlay               off                            default
tank/global1/root                  encryption            off                            default
tank/global1/root                  keylocation           none                           default
tank/global1/root                  keyformat             none                           default
tank/global1/root                  pbkdf2iters           0                              default
tank/global1/root                  special_small_blocks  0                              default
```

**Comment by deliciouslytyped** [comment]
I did some further testing and I don't seem to have a problem with a dataset that is not my root dataset, however I'm not sure yet if I tested properly.

**Comment by deliciouslytyped** [comment]
My `/` is mounted like:
```
tank/global1/root on / type zfs (rw,relatime,xattr,noacl)
tank/global1/root on /nix/store type zfs (ro,relatime,xattr,noacl)
```
My test device:
zfs list:
```
tank/global1/testd            1.19G   116G     1.19G  legacy
```
mount:
```
bakery4/testd on /run/user/1000/tmp.VyocmCihuu/mnt type zfs (rw,relatime,xattr,noacl)
```

**Comment by deliciouslytyped** [comment]
Here are the pool settings: 
```
tank  size                           216G                           -
tank  capacity                       43%                            -
tank  altroot                        -                              default
tank  health                         ONLINE                         -
tank  guid                           [censored]                     -
tank  version                        -                              default
tank  bootfs                         -                              default
tank  delegation                     on                             default
tank  autoreplace                    off                            default
tank  cachefile                      -                              default
tank  failmode                       wait                           default
tank  listsnapshots                  off                            default
tank  autoexpand                     off                            default
tank  dedupditto                     0                              default
tank  dedupratio                     1.00x                          -
tank  free                           122G                           -
tank  allocated                      93.7G                          -
tank  readonly                       off                            -
tank  ashift                         12                             local
tank  comment                        -                              default
tank  expandsize                     720G                           -
tank  freeing                        0                              -
tank  fragmentation                  11%                            -
tank  leaked                         0                              -
tank  multihost                      off                            default
tank  checkpoint                     -                              -
tank  load_guid                      [censored]                     -
tank  autotrim                       off                            default
tank  feature@async_destroy          enabled                        local
tank  feature@empty_bpobj            active                         local
tank  feature@lz4_compress           active                         local
tank  feature@multi_vdev_crash_dump  enabled                        local
tank  feature@spacemap_histogram     active                         local
tank  feature@enabled_txg            active                         local
tank  feature@hole_birth             active                         local
tank  feature@extensible_dataset     active                         local
tank  feature@embedded_data          active                         local
tank  feature@bookmarks              enabled                        local
tank  feature@filesystem_limits      enabled                        local
tank  feature@large_blocks           enabled                        local
tank  feature@large_dnode            enabled                        local
tank  feature@sha512                 enabled                        local
tank  feature@skein                  enabled                        local
tank  feature@edonr                  enabled                        local
tank  feature@userobj_accounting     active                         local
tank  feature@encryption             enabled                        local
tank  feature@project_quota          active                         local
tank  feature@device_removal         enabled                        local
tank  feature@obsolete_counts        enabled                        local
tank  feature@zpool_checkpoint       enabled                        local
tank  feature@spacemap_v2            active                         local
tank  feature@allocation_classes     enabled                        local
tank  feature@resilver_defer         enabled                        local
tank  feature@bookmark_v2            enabled                        local
```

**Comment by deliciouslytyped** [comment]
I also did a sanity check by running `cat /dev/urandom | pv | dd if=/dev/stdin of=./test` on a tmpfs as well, and the system continued to function properly.

**Comment by deliciouslytyped** [comment]
I happen to have an extra unformatted partition (sda2 above), and tested `cat /dev/urandom | pv -B1M | sudo dd if=/dev/stdin of=/dev/sda2` on it. The throughput seems to max at approximately 140MB/s, however it has higher than ZFS, and goes down to 40MB/s or so if I send an IRC message (I did not test this thoroughly). This should prove there is nothing wrong with the underlying hardware device.

This seems to point towards ZFS having some sort of bad IO scheduling behaviour?

**Comment by deliciouslytyped** [comment]
I'm not sure what I did, but `zfs recv` seems to limit things to about 30MB/s right now and its jumping between that and 10MB/s.
